File: src/elf_loader.c
#define _GNU_SOURCE

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <errno.h>
#include <unistd.h>
#include <fcntl.h>
#include <sys/mman.h>
#include <sys/syscall.h>
#include <elf.h>

#include "elf_loader.h"

#define PAGE_SIZE 4096UL

#define PAGE_DOWN(x) ((x) &~(PAGE_SIZE - 1)) //Round an address DOWN to the nearest page boundary
#define PAGE_UP(x) PAGE_DOWN((x) + PAGE_SIZE - 1) //Round an address UP to the next page boundary
#define PUSH_VAL(v, loader)  do { \
    if (loader->is_64bit) { \
        *(uint64_t *)p = (uint64_t)(uintptr_t)(v); \
    } else { \
        *(uint32_t *)p = (uint32_t)(uintptr_t)(v); \
    } \
    p += pw; \
} while(0)

#if defined(__x86_64__)

static void __attribute__((noreturn))
jump_to_entry(void *sp, void *entry)
{
    __asm__ volatile(
        "mov  %0, %%rsp\n\t"
        "xor  %%rbp, %%rbp\n\t"
        "xor  %%rdx, %%rdx\n\t"
        "jmp  *%1\n\t"
        : : "r"(sp), "r"(entry) : "memory"
    );

    __builtin_unreachable();
}

#elif defined(__i386__)

static void __attribute__((noreturn))
jump_to_entry(void *sp, void *entry)
{
    __asm__ volatile(
        "mov  %0, %%esp\n\t"
        "xor  %%ebp, %%ebp\n\t"
        "xor  %%edx, %%edx\n\t"
        "jmp  *%1\n\t"
        : : "r"(sp), "r"(entry) : "memory");
    __builtin_unreachable();
}

#elif defined(__aarch64__)

static void __attribute__((noreturn))
jump_to_entry(void *sp, void *entry)
{
    __asm__ volatile(
        "mov  sp,  %0\n\t"
        "mov  x29, #0\n\t"
        "mov  x30, #0\n\t"
        "br   %1\n\t"
        : : "r"(sp), "r"(entry) : "memory");
    __builtin_unreachable();
}

#elif defined(__arm__)

static void __attribute__((noreturn))
jump_to_entry(void *sp, void *entry)
{
    __asm__ volatile(
        "mov  sp,  %0\n\t"
        "mov  fp,  #0\n\t"
        "mov  lr,  #0\n\t"
        "bx   %1\n\t"
        : : "r"(sp), "r"(entry) : "memory");
    __builtin_unreachable();
}

#elif defined(__riscv)
static void __attribute__((noreturn))
jump_to_entry(void *sp, void *entry)
{
    __asm__ volatile(
        "mv   sp,  %0\n\t"
        "li   fp,  0\n\t"
        "li   ra,  0\n\t"
        "jr   %1\n\t"
        : : "r"(sp), "r"(entry) : "memory");
    __builtin_unreachable();
}

#else

#error "No jump_to_entry() trampoline for this architecture."

#endif

int protect_from_flags(uint32_t f)
{
    return ((f & PF_R) ? PROT_READ : 0) | ((f & PF_W) ? PROT_WRITE : 0) | ((f & PF_X) ? PROT_EXEC : 0);
}

int elf_probe(const void *data, size_t size, ElfLoader *out)
{
    memset(out, 0, sizeof(*out));

    ElfHeader hdr;
    if (elf_parse_header(data, size, &hdr) != 0)
    {
        fprintf(stderr, "Invalid ELF file\n");
        return -1;
    }

    out->arch = hdr.arch;
    out->is_64bit = hdr.is_64bit;
    out->is_pie = (hdr.e_type == ET_DYN);

    ElfPhdr phdr;
    for (int i = 0; i < hdr.e_phnum; i++)
    {
        if (elf_get_phdr(data, size, &hdr, i, &phdr) == 0 && phdr.p_type == PT_INTERP)
        {
            out->interp_offset = (uintptr_t)phdr.p_offset;
            break;
        }
    }

    return 0;
}

int elf_load(const void *data, size_t size, ElfLoader *out)
{
    if (elf_probe(data, size, out) != 0)
        return -1;

    if (!elf_is_native(out->arch))
    {
        fprintf(stderr, "Unmatched: Binary is %s but host is %s\n", elf_arch_name(out->arch), ELF_HOST_NAME);
        return -1;
    }

    if (out->interp_offset)
    {
        const char *interp = (const char *)data + out->interp_offset;
        fprintf(stderr, "This binary requires dynamic linker '%s'.\nPlease use elf_memfd_exec() for dynamic binary.\n", interp);
    }

    ElfHeader hdr;
    elf_parse_header(data, size, &hdr);

    uintptr_t vmin = UINTPTR_MAX;
    uintptr_t vmax = 0;
    ElfPhdr phdr;
    for (int i = 0; i < hdr.e_phnum; i++)
    {
        if (elf_get_phdr(data, size, &hdr, i, &phdr) != 0)
            continue;

        if (phdr.p_type != PT_LOAD)
            continue;

        uintptr_t s = PAGE_DOWN(phdr.p_vaddr); //segment start
        uintptr_t e = PAGE_UP(phdr.p_vaddr + phdr.p_memsize); //segment end

        vmin = s < vmin ? s : vmin;
        vmax = e > vmax ? e : vmax;
    }

    if (UINTPTR_MAX == vmin)
    {
        fprintf(stderr, "PT_LOAD segments not found\n");
        return -1;
    }

    size_t map_size = vmax - vmin;
    int is_pie = out->is_pie;

    void *base = mmap(
        is_pie ? NULL : (void *)vmin, 
        map_size, 
        PROT_NONE, 
        MAP_PRIVATE | MAP_ANONYMOUS | (is_pie ? 0 : MAP_FIXED), 
        -1, 
        0
    );
    if (MAP_FAILED == base)
    {
        perror("mmap: reserve virtual range");
        return -1;
    }

    uintptr_t bias = is_pie ? (uintptr_t)base - vmin : 0;
    out->base = base;
    out->map_size = map_size;
    out->bias = bias;

    for (int i = 0; i < hdr.e_phnum; i++)
    {
        if (elf_get_phdr(data, size, &hdr, i, &phdr) != 0)
            continue;

        if (PT_LOAD != phdr.p_type)
            continue;

        uintptr_t segment_start = PAGE_DOWN(phdr.p_vaddr + bias);
        uintptr_t segment_end = PAGE_UP(phdr.p_vaddr + bias + phdr.p_memsize);
        size_t segment_length = segment_end - segment_start;

        void *mem = mmap(
            (void *)segment_start, 
            segment_length, 
            PROT_READ | PROT_WRITE,
            MAP_PRIVATE | MAP_ANONYMOUS | MAP_FIXED,
            -1,
            0
        );

        if (MAP_FAILED == mem)
        {
            perror("mmap: mmap error");
            goto failure;
        }

        size_t copy_bytes = phdr.p_filesize;
        if (phdr.p_offset + copy_bytes > size)
            copy_bytes = size - phdr.p_offset;

        memcpy((char *)(phdr.p_vaddr + bias), (const char *)data + phdr.p_offset, copy_bytes);

        if (phdr.p_memsize > phdr.p_filesize)
            memset((char *)(phdr.p_vaddr + bias) + phdr.p_filesize, 0, phdr.p_memsize - phdr.p_filesize);

        if (mprotect((void *)segment_start, segment_length, protect_from_flags(phdr.p_flags)) != 0)
        {
            perror("mprotect: permissions error");
            goto failure;
        }
    }

    out->entry = (void *)(hdr.e_entry + bias);

    size_t stack_size = ELF_DEFAULT_STACK_SIZE;
    void *stack = mmap(NULL, stack_size, PROT_READ | PROT_WRITE, MAP_PRIVATE | MAP_ANONYMOUS | MAP_GROWSDOWN, -1, 0);
    if (MAP_FAILED == stack)
    {
        perror("mmap: stack error");
        goto failure;
    }

    out->stack_base = stack;
    out->stack_size = stack_size;

    fprintf(stdout, "Loaded %s ELF%d at %p, entry=%p, bias=0x%lx\n", 
        elf_arch_name(out->arch),
        out->is_64bit ? 64 : 32,
        base,
        out->entry,
        (unsigned long)bias
    );

    return 0;

failure:
    munmap(base, map_size);
    return -1;
}

void elf_execute(const ElfLoader *loader, int argc, char **argv, char **envp)
{
    int envc = 0;
    while (envp && envp[envc])
        envc++;

    char *stk = (char *)loader->stack_base + loader->stack_size;
    stk = (char *)((uintptr_t)stk & ~15UL);

    char **new_envp = malloc((envc + 1) * sizeof(char *));
    char **new_argv = malloc((argc + 1) * sizeof(char *));

    for (int i = envc - 1; i >= 0; i--)
    {
        size_t len = strlen(envp[i]) + 1;
        stk -= len;
        memcpy(stk, envp[i], len);
        new_envp[i] = stk;
    }

    new_envp[envc] = NULL;

    for (int i = argc - 1; i >= 0; i--)
    {
        size_t len = strlen(argv[i]) + 1;
        stk -= len;
        memcpy(stk, argv[i], len);
        new_argv[i] = stk;
    }

    new_argv[argc] = NULL;

    stk = (char *)((uintptr_t)stk & ~15UL);

    Aux64 auxv[] =
    {
        { AT_PAGESZ, PAGE_SIZE },
        { AT_BASE, (uintptr_t)loader->base },
        { AT_FLAGS, 0 },
        { AT_ENTRY, (uintptr_t)loader->entry },
        { AT_UID, (uint64_t)getuid() },
        { AT_EUID, (uint64_t)geteuid() },
        { AT_GID, (uint64_t)getgid() },
        { AT_EGID, (uint64_t)getegid() },
        { AT_SECURE, 0 },
        { AT_NULL, 0 },
    };

    int n_aux = (int)(sizeof(auxv) / sizeof(auxv[0]));
    int pw = loader->is_64bit ? 8 : 4;
    int n_slots = 1 + argc + 1 + envc + 1 + n_aux * 2;
    stk -= (size_t)n_slots * (size_t)pw;
    stk = (char *)((uintptr_t)stk & ~15UL);

    char *p = stk;
    PUSH_VAL(argc, loader);
    for (int i = 0; i < argc; i++)
        PUSH_VAL(new_argv[i], loader);
    PUSH_VAL(0, loader);

    for (int i = 0; i < envc; i++)
        PUSH_VAL(new_envp[i], loader);
    PUSH_VAL(0, loader);

    for (int i = 0; i < n_aux; i++)
    {
        PUSH_VAL(auxv[i].type, loader);
        PUSH_VAL(auxv[i].val, loader);
    }

    free(new_argv);
    free(new_envp);

    fprintf(stdout, "%s: entry=%p sp=%p\n", elf_arch_name(loader->arch), loader->entry, stk);

    jump_to_entry(stk, loader->entry);
}

void elf_unload(ElfLoader *loader)
{
    if (loader->base)
        munmap(loader->base, loader->map_size);
    
    if (loader->stack_base)
        munmap(loader->stack_base, loader->stack_size);

    memset(loader, 0, sizeof(*loader));
}

void elf_info(const ElfLoader *loader)
{
    fprintf(stdout,
        "ELF Info:\n"
        "\tArchitecture: %s (%s)\n"
        "\tType: %s\n"
        "\tMapped base: %p\n"
        "\tLoad bias: 0x%lx\n"
        "\tEntry point: %p\n"
        "\tInterp: %s\n"
        "\tNative exec: %s\n",

        elf_arch_name(loader->arch),
        loader->is_64bit ? "ELF64" : "ELF32",
        loader->is_pie ? "PIE (ET_DYN)" : "static (ET_EXEC)",
        loader->base,
        (unsigned long)loader->bias,
        loader->entry,
        loader->interp_offset ? "YES (dynamic)" : "NO (static)",
        elf_is_native(loader->arch) ? "YES" : "NO (cross-arch)"
    );
}

//Execute ELF via memory file descriptor.
int elf_memfd_exec(const void *data, size_t size, int argc, char **argv, char **envp)
{
#ifdef __NR_memfd_create
    /*  Not all Linux systems have this syscall.
        Create an anonymous in-memory file.
    */
    int mfd = (int)syscall(__NR_memfd_create, "elf_mem", 0);
    if (mfd < 0)
    {
        //syscall failed.
        perror("memfd_create");
        return -1;
    }

    /*  Write the entire ELF image into the memfd.
        Using loop because write() may return fewer bytes than requested for Linux systems.
    */ 
    size_t done = 0;
    while (done < size)
    {
        ssize_t n = write(mfd, (unsigned char *)data + done, size - done);
        if (n <= 0)
        {
            //Write error.
            perror("Write into memfd failed.");
            close(mfd);

            return -1;
        }

        done += (size_t)n; //Absoultely positive.
    }

    /*  Directly execute the ELF image from the file descriptor.
        The kernel's ELF binfmt handler takes over from here.
    */
   fexecve(mfd, argv, envp);

   //If it reaches here, fexecve failed.
   perror("fexecve() failed.");
   close(mfd);

   return -1;

#else
    //Kernel doesn't have memfd_create (pre 3.17 or non-Linux)
    fprintf(stderr, "memfd_create (__NR_memfd_create) not available on this kernel.");
    
    //Avoiding warning, learn from: https://gcc.gnu.org/onlinedocs/gcc/Warning-Options.html
    (void)data; (void)size; (void)argc; (void)argv; (void)envp;
    
    return -1;

#endif
}

File: Rust/src/elf_loader.rs
//elf_loader.rs

use std::ptr;
use libc::{
    close, fexecve, mmap, mprotect, munmap, write,
    MAP_ANON, MAP_ANONYMOUS, MAP_FAILED, MAP_FIXED, MAP_GROWSDOWN, MAP_PRIVATE, PROT_NONE, PROT_READ, PROT_WRITE,
};

use crate::elf_arch::{
    ElfArch, ElfHeader, ElfPhdr,
    elf_get_phdr, elf_parse_header, prot_from_flags, host_arch_name,
    ET_DYN, PT_LOAD, PT_INTERP, ELF_DEFAULT_STACK_SIZE,
};

const PAGE_SIZE: usize = 4096;

#[inline]
fn page_down(x: usize) -> usize {
    return x & !(PAGE_SIZE - 1);
}

#[inline]
fn page_up(x: usize) -> usize {
    return page_down(x + PAGE_SIZE - 1);
}

#[derive(Debug)]
pub struct ElfLoader {
    pub arch         : ElfArch,   // architecture
    pub is_64bit     : bool,      // is 64-bit
    pub is_pie       : bool,      // is Position-Independent Executable (PIE, for ASLR)
    pub interp_offset: usize,     // interpreter offset
    pub base         : *mut u8,   // base address
    pub map_size     : usize,     // map size
    pub load_bias    : usize,     // bias
    pub entry        : *mut u8,   // entry point
    pub stack_base   : *mut u8,   // base address of stack
    pub stack_size   : usize,     // stack size

    // stored for auxv AT_PHDR/AT_PHENT/AT_PHNUM
    // libc _start (musl, glibc) reads these to find the program headers
    pub phdr_offset  : usize,     // e_phoff: file offset of program header table
    pub phdr_entsize : usize,     // e_phentsize: size of one program header entry
    pub phdr_num     : usize,     // e_phnum: number of program header entries
}

impl Default for ElfLoader {
    fn default() -> Self {
        Self {
            arch         : ElfArch::Unknown,
            is_64bit     : false,
            is_pie       : false,
            interp_offset: 0,
            base         : ptr::null_mut(),
            map_size     : 0,
            load_bias    : 0,
            entry        : ptr::null_mut(),
            stack_base   : ptr::null_mut(),
            stack_size   : 0,
            phdr_offset  : 0,
            phdr_entsize : 0,
            phdr_num     : 0,
        }
    }
}

// probe ELF
pub fn elf_probe(data: &[u8]) -> Result<ElfLoader, String> {
    let hdr = elf_parse_header(data).ok_or_else(|| "Invalid ELF file".to_string())?;
    let mut loader = ElfLoader::default();

    loader.arch     = hdr.arch;
    loader.is_64bit = hdr.is_64bit;
    loader.is_pie   = hdr.e_type == ET_DYN;

    for i in 0..hdr.e_phnum {
        if let Some(ph) = elf_get_phdr(data, &hdr, i) {
            if ph.p_type == PT_INTERP {
                loader.interp_offset = ph.p_offset as usize;
                break;
            }
        }
    }

    return Ok(loader);
}

// read interpreter path
fn interp_path(data: &[u8], offset: usize) -> String {
    if offset >= data.len() {
        return "Invalid".to_string();
    }

    let slice = &data[offset..];
    let p = slice.iter().position(|&b| b == 0).unwrap_or(slice.len());

    return String::from_utf8_lossy(&slice[..p]).to_string();
}

fn map_segment(data: &[u8], ph: &ElfPhdr, bias: usize) -> Result<(), String> {
    let seg_start = page_down(ph.p_vaddr as usize + bias);
    let seg_end   = page_up(ph.p_vaddr as usize + bias + ph.p_memsize as usize);
    let seg_len   = seg_end - seg_start;

    let mem = unsafe {
        mmap(
            seg_start as *mut libc::c_void,
            seg_len,
            PROT_READ | PROT_WRITE,
            MAP_PRIVATE | MAP_ANON | MAP_FIXED,
            -1,
            0,
        )
    };

    if MAP_FAILED == mem {
        return Err(format!("mmap segment failed: {}", std::io::Error::last_os_error()));
    }

    let dest       = (ph.p_vaddr as usize + bias) as *mut u8;
    let src_offset = ph.p_offset as usize;
    let copy_bytes = (ph.p_filesize as usize).min(data.len() - src_offset);

    unsafe {
        ptr::copy_nonoverlapping(data.as_ptr().add(src_offset), dest, copy_bytes);
    }

    // Zero the BSS tail (p_memsize > p_filesize region)
    if ph.p_memsize > ph.p_filesize {
        let bss_start = (ph.p_vaddr as usize + bias + ph.p_filesize as usize) as *mut u8;
        let bss_len   = (ph.p_memsize - ph.p_filesize) as usize;
        unsafe { ptr::write_bytes(bss_start, 0, bss_len); }
    }

    let prot = prot_from_flags(ph.p_flags);
    if unsafe { mprotect(seg_start as *mut libc::c_void, seg_len, prot) } != 0 {
        return Err(format!("mprotect failed: {}", std::io::Error::last_os_error()));
    }

    return Ok(());
}

pub fn elf_load(data: &[u8]) -> Result<ElfLoader, String> {
    let mut loader = elf_probe(data)?;

    if !loader.arch.is_native() {
        return Err(format!("Cross-arch: binary is {} host is {}", loader.arch.to_str(), host_arch_name()));
    }

    if loader.interp_offset != 0 {
        let interp = interp_path(data, loader.interp_offset);
        eprintln!("Warning: Binary requires dynamic linker '{}'.\nUse elf_memfd_exec() for dynamic binary.", interp);
    }

    let hdr = elf_parse_header(data).unwrap();

    // Store program header metadata for auxv AT_PHDR/AT_PHENT/AT_PHNUM.
    // libc's _start reads these to locate the program header table.
    loader.phdr_offset  = hdr.e_phoffset  as usize;
    loader.phdr_entsize = hdr.e_phentsize as usize;
    loader.phdr_num     = hdr.e_phnum     as usize;

    let mut vmin = usize::MAX;
    let mut vmax = 0usize;

    for i in 0..hdr.e_phnum {
        let ph = match elf_get_phdr(data, &hdr, i) {
            Some(p) if p.p_type == PT_LOAD => p,
            _ => continue,
        };

        let start = page_down(ph.p_vaddr as usize);
        let end   = page_up(ph.p_vaddr as usize + ph.p_memsize as usize);

        if start < vmin { vmin = start; }
        if end   > vmax { vmax = end;   }
    }

    if usize::MAX == vmin {
        return Err("No PT_LOAD segments found".to_string());
    }

    let map_size = vmax - vmin;
    let is_pie   = loader.is_pie;

    // ---- STEP 1: allocate stack FIRST ----
    // Stack must be allocated before reserving the ELF virtual range.
    // If we reserve ELF range first, the kernel may place the stack inside
    // that reservation (especially for PIE where the kernel picks the base freely).
    // Stack overlapping with code/data = corrupted sp = segfault.
    let stack = unsafe {
        mmap(
            ptr::null_mut(),
            ELF_DEFAULT_STACK_SIZE,
            PROT_READ | PROT_WRITE,
            MAP_PRIVATE | MAP_ANONYMOUS | MAP_GROWSDOWN,
            -1,
            0,
        )
    };

    if MAP_FAILED == stack {
        return Err(format!("mmap stack failed: {}", std::io::Error::last_os_error()));
    }

    loader.stack_base = stack as *mut u8;
    loader.stack_size = ELF_DEFAULT_STACK_SIZE;

    // ---- STEP 2: reserve the full ELF virtual range with PROT_NONE ----
    // Stack is already claimed so the kernel won't overlap it with ELF base.
    let hint  = if is_pie { ptr::null_mut() } else { vmin as *mut libc::c_void };
    let flags = MAP_PRIVATE | MAP_ANONYMOUS | if is_pie { 0 } else { MAP_FIXED };

    let base = unsafe {
        mmap(hint, map_size, PROT_NONE, flags, -1, 0)
    };

    if MAP_FAILED == base {
        unsafe { munmap(stack, ELF_DEFAULT_STACK_SIZE); }
        return Err(format!("mmap reservation failed: {}", std::io::Error::last_os_error()));
    }

    let base      = base as *mut u8;
    let load_bias = if is_pie { base as usize - vmin } else { 0 };

    loader.base      = base;
    loader.map_size  = map_size;
    loader.load_bias = load_bias;

    // ---- STEP 3: map each PT_LOAD segment over the reservation ----
    for i in 0..hdr.e_phnum {
        let ph = match elf_get_phdr(data, &hdr, i) {
            Some(p) if PT_LOAD == p.p_type => p,
            _ => continue,
        };

        if let Err(e) = map_segment(data, &ph, load_bias) {
            // Clean up both reservations on failure
            unsafe {
                munmap(base as *mut libc::c_void, map_size);
                munmap(stack, ELF_DEFAULT_STACK_SIZE);
            }
            return Err(e);
        }
    }

    loader.entry = (hdr.e_entry as usize + load_bias) as *mut u8;

    eprintln!("Loaded {} ELF{} at {:p}, entry={:p}, bias=0x{:x}",
        loader.arch.to_str(),
        if loader.is_64bit { 64 } else { 32 },
        loader.base,
        loader.entry,
        loader.load_bias,
    );

    return Ok(loader);
}

// ------------------------------------------------------------------ //
// Execution                                                          //
// ------------------------------------------------------------------ //

#[cfg(target_arch = "x86_64")]
unsafe fn jump_to_entry(sp: *mut u8, entry: *mut u8) -> ! {
    std::arch::asm!(
        "mov rsp, {sp}",
        "xor rbp, rbp",
        "xor rdx, rdx",
        "jmp {entry}",
        sp    = in(reg) sp,
        entry = in(reg) entry,
        options(nostack, noreturn)
    );
}

#[cfg(target_arch = "x86")]
unsafe fn jump_to_entry(sp: *mut u8, entry: *mut u8) -> ! {
    std::arch::asm!(
        "mov esp, {sp}",
        "xor ebp, ebp",
        "xor edx, edx",
        "jmp {entry}",
        sp    = in(reg) sp,
        entry = in(reg) entry,
        options(nostack, noreturn)
    );
}

#[cfg(target_arch = "aarch64")]
unsafe fn jump_to_entry(sp: *mut u8, entry: *mut u8) -> ! {
    std::arch::asm!(
        "mov sp,  {sp}",
        "mov x29, #0",
        "mov x30, #0",
        "br  {entry}",
        sp    = in(reg) sp,
        entry = in(reg) entry,
        options(nostack, noreturn)
    );
}

#[cfg(target_arch = "arm")]
unsafe fn jump_to_entry(sp: *mut u8, entry: *mut u8) -> ! {
    std::arch::asm!(
        "mov sp, {sp}",
        "mov fp, #0",
        "mov lr, #0",
        "bx  {entry}",
        sp    = in(reg) sp,
        entry = in(reg) entry,
        options(nostack, noreturn)
    );
}

#[cfg(target_arch = "riscv64")]
unsafe fn jump_to_entry(sp: *mut u8, entry: *mut u8) -> ! {
    std::arch::asm!(
        "mv  sp, {sp}",
        "li  fp, 0",
        "li  ra, 0",
        "jr  {entry}",
        sp    = in(reg) sp,
        entry = in(reg) entry,
        options(nostack, noreturn)
    );
}

#[cfg(not(any(
    target_arch = "x86_64",
    target_arch = "x86",
    target_arch = "aarch64",
    target_arch = "arm",
    target_arch = "riscv64",
)))]
unsafe fn jump_to_entry(_sp: *mut u8, _entry: *mut u8) -> ! {
    panic!("jump_to_entry: unsupported architecture");
}

pub unsafe fn elf_execute(loader: &ElfLoader, args: &[String], env: &[String]) -> ! {
    let mut stk = (loader.stack_base as usize + loader.stack_size) & !15usize;

    // ---- push env strings onto the stack (high to low) ----
    let mut env_ptrs: Vec<*const u8> = Vec::with_capacity(env.len() + 1);
    for s in env.iter().rev() {
        let bytes = s.as_bytes();
        stk -= bytes.len() + 1;
        let dst = stk as *mut u8;
        ptr::copy_nonoverlapping(bytes.as_ptr(), dst, bytes.len());
        *dst.add(bytes.len()) = 0;
        env_ptrs.push(dst);
    }
    env_ptrs.reverse();
    env_ptrs.push(ptr::null());

    // ---- push arg strings onto the stack (high to low) ----
    let mut argu_ptrs: Vec<*const u8> = Vec::with_capacity(args.len() + 1);
    for s in args.iter().rev() {
        let bytes = s.as_bytes();
        stk -= bytes.len() + 1;
        let dst = stk as *mut u8;
        ptr::copy_nonoverlapping(bytes.as_ptr(), dst, bytes.len());
        *dst.add(bytes.len()) = 0;
        argu_ptrs.push(dst);
    }
    argu_ptrs.reverse();
    argu_ptrs.push(ptr::null());

    stk &= !15usize;

    // AT_PHDR = base + phdr_offset: the in-memory address of the program
    // header table. libc's _start reads this to find TLS segments, vDSO, etc.
    let at_phdr = loader.base as usize + loader.phdr_offset;

    let auxv: &[(usize, usize)] = &[
        (libc::AT_PAGESZ as usize, PAGE_SIZE),
        (libc::AT_BASE   as usize, loader.base  as usize),
        (libc::AT_FLAGS  as usize, 0),
        (libc::AT_ENTRY  as usize, loader.entry as usize),
        // program header info — required by musl/glibc _start
        (libc::AT_PHDR   as usize, at_phdr),
        (libc::AT_PHENT  as usize, loader.phdr_entsize),
        (libc::AT_PHNUM  as usize, loader.phdr_num),
        (libc::AT_UID    as usize, libc::getuid()  as usize),
        (libc::AT_EUID   as usize, libc::geteuid() as usize),
        (libc::AT_GID    as usize, libc::getgid()  as usize),
        (libc::AT_EGID   as usize, libc::getegid() as usize),
        (libc::AT_SECURE as usize, 0),
        (0, 0), // AT_NULL terminator
    ];

    let pw = if loader.is_64bit { 8usize } else { 4usize };

    let n_slots = 1
        + args.len() + 1
        + env.len()  + 1
        + auxv.len() * 2;

    stk -= n_slots * pw;
    stk &= !15usize;

    let mut p = stk;

    macro_rules! push_val {
        ($v:expr) => {
            if loader.is_64bit {
                *(p as *mut u64) = $v as u64;
            } else {
                *(p as *mut u32) = $v as u32;
            }
            p += pw;
        };
    }

    push_val!(args.len());
    for ptr in &argu_ptrs { push_val!(*ptr as usize); }
    for ptr in &env_ptrs  { push_val!(*ptr as usize); }
    for (t, v) in auxv {
        push_val!(*t);
        push_val!(*v);
    }

    eprintln!("{}: entry={:p} sp=0x{:x}", loader.arch.to_str(), loader.entry, stk);

    jump_to_entry(stk as *mut u8, loader.entry);
}

pub fn elf_unload(loader: &mut ElfLoader) {
    unsafe {
        if !loader.base.is_null() {
            munmap(loader.base as *mut libc::c_void, loader.map_size);
            loader.base = ptr::null_mut();
        }

        if !loader.stack_base.is_null() {
            munmap(loader.stack_base as *mut libc::c_void, loader.stack_size);
            loader.stack_base = ptr::null_mut();
        }

        *loader = ElfLoader::default();
    }
}

pub fn elf_print_info(loader: &ElfLoader) {
    println!(
        "ELF Info:\n\
        \tArchitecture : {} ({})\n\
        \tType         : {}\n\
        \tMapped base  : {:p}\n\
        \tLoad bias    : 0x{:x}\n\
        \tEntry point  : {:p}\n\
        \tInterp       : {}\n\
        \tNative exec  : {}",
        loader.arch.to_str(),
        if loader.is_64bit { "ELF64" } else { "ELF32" },
        if loader.is_pie { "PIE (ET_DYN)" } else { "static (ET_EXEC)" },
        loader.base,
        loader.load_bias,
        loader.entry,
        if loader.interp_offset != 0 { "YES (dynamic)" } else { "NO (static)" },
        if loader.arch.is_native()   { "YES" }           else { "NO (cross-arch)" },
    );
}

fn to_cstring_vec(strings: &[String]) -> Vec<std::ffi::CString> {
    return strings.iter()
        .map(|s| std::ffi::CString::new(s.as_str()).unwrap_or_default())
        .collect();
}

pub fn elf_memfd_exec(data: &[u8], args: &[String], env: &[String]) -> Result<(), String> {
    let c_args = to_cstring_vec(args);
    let c_env  = to_cstring_vec(env);
    let argv: Vec<*const libc::c_char> = c_args.iter()
        .map(|s| s.as_ptr()).chain(std::iter::once(ptr::null())).collect();
    let envp: Vec<*const libc::c_char> = c_env.iter()
        .map(|s| s.as_ptr()).chain(std::iter::once(ptr::null())).collect();

    unsafe {
        let mfd = libc::syscall(libc::SYS_memfd_create, b"elf_mem\0".as_ptr(), 0u32) as i32;
        if mfd < 0 {
            return Err(format!("memfd_create failed: {}", std::io::Error::last_os_error()));
        }

        let mut done = 0usize;
        while done < data.len() {
            let n = write(mfd, data.as_ptr().add(done) as *const libc::c_void, data.len() - done);
            if n <= 0 {
                close(mfd);
                return Err(format!("write to memfd failed: {}", std::io::Error::last_os_error()));
            }
            done += n as usize;
        }

        fexecve(mfd, argv.as_ptr(), envp.as_ptr());

        let err = std::io::Error::last_os_error();
        close(mfd);
        return Err(format!("fexecve failed: {}", err));
    }
}